### PR TITLE
Initial codebase

### PR DESCRIPTION
**Pull request by gene1wood**

### CODE REVIEW

File: functions/federated_aws_rp/app.py [hunk 1]
```diff
@@ -0,0 +1,248 @@
...
+        return {
+            'statusCode': 403,
+            'headers': {'Content-Type': 'text/html'},
+            'body': "Access denied : {}".format(e)}
+    token = exchange_code_for_token(
+        store['code_verifier'],
+        discovery_document['token_endpoint'],
+        code)
+    jwt.decode(
+        token=token["id_token"],
```
Comment (april) [inline functions/federated_aws_rp/app.py]: There is some mixing of quotation mark styles in here.  :)

File: functions/federated_aws_rp/app.py [hunk 2]
```diff
@@ -0,0 +1,248 @@
...
+    global discovery_document
+    if 'discovery_document' not in globals():
+        logger.debug('Fetching discovery document')
+        discovery_document = get_discovery_document(CONFIG.discovery_url)
+
+    try:
+        if 'state' is None:
+            raise AccessDenied('"state" query parameter is missing"')
+        store = get_store(cookie_header)
+        if store.get('state') != state:
```
Comment (april) [inline functions/federated_aws_rp/app.py:80]: `elif`s, maybe?

**Comment by gene1wood** [comment]
Here's [a video of the refresh functionality in play](https://youtu.be/qKPD_xzeCTo)

File: functions/federated_aws_rp/app.py [hunk 3]
```diff
@@ -0,0 +1,248 @@
...
+    :param role_arn: The AWS IAM Role ARN selected from the role picker
+    :return: An AWS API Gateway output dictionary for proxy mode
+    """
+    global discovery_document
+    if 'discovery_document' not in globals():
+        logger.debug('Fetching discovery document')
+        discovery_document = get_discovery_document(CONFIG.discovery_url)
+
+    store = get_store(cookie_header)
+    store['role_arn'] = role_arn
```
Comment (april) [inline functions/federated_aws_rp/app.py:46]: You could maybe use `extend` here, but it's probably not a big deal when there's just one additional thing stuffed in there.

File: functions/federated_aws_rp/app.py [hunk 4]
```diff
@@ -0,0 +1,248 @@
...
+                }
+            )
+
+        # Sort the list by role name
+        roles["roles"] = sorted(roles["roles"], key=lambda x: x['name'])
+        response = {
+            'statusCode': 200,
+            'headers': {
+                'Content-Type': 'text/html',
+                'Set-Cookie': '{}={}; Secure; HttpOnly; Max-Age=3600'.format(
```
Comment (april) [inline functions/federated_aws_rp/app.py]: I think you can probably use `SameSite` here, to protect you against CSRF attacks.

File: functions/federated_aws_rp/config.py [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+import urllib.parse
+
+from jinja2 import Template
+
+
+class Config:
+    pass
+
+
+CONFIG = Config()
```
Comment (april) [inline functions/federated_aws_rp/config.py]: This all seems a bit randomly ordered. Maybe set it inside `__init__` and order the variables a bit?

File: functions/federated_aws_rp/utils.py [hunk 6]
```diff
@@ -0,0 +1,298 @@
...
+    """Fetch the OIDC discovery document and enrich it with the JWKS
+
+    Fetch the the discovery_url and parse it into a dictionary. Fetch the
+    content of the jwks_uri and add it to the dictionary in the "jkws" key
+
+    :param discovery_url: The URL of the identity providers OIDC discovery url
+    :return: A dictionary containing information about the identity provider
+    """
+    result = requests.get(discovery_url).json()
+    result['jwks'] = requests.get(result['jwks_uri']).json()
```
Comment (april) [inline functions/federated_aws_rp/utils.py:74]: Just a note that this all has the same `requests` stuff we currently have with not handling exceptions.

File: functions/federated_aws_rp/utils.py [hunk 7]
```diff
@@ -0,0 +1,298 @@
...
+    }
+
+    # TODO : Wait should we not do PKCE because this is a traditional RP
+    # and instead use a client secret?
+    logger.debug('Requesting token with arguments {}'.format(body))
+    token = requests.post(
+        token_endpoint,
+        headers=headers,
+        json=body).json()
+    return token
```
Comment (april) [inline functions/federated_aws_rp/utils.py]: I would consolidate these previous two statements.

**Review by april — CHANGES_REQUESTED** [review]

File: functions/federated_aws_rp/app.py (hunk 3 above)
Comment (gene1wood) [reply functions/federated_aws_rp/app.py:46]: My goal with this was to take the `store` dict and add a new key value pair to it. I think [`extend`](https://docs.python.org/3/tutorial/datastructures.html#more-on-lists) is for lists though not dicts?

File: functions/federated_aws_rp/app.py (hunk 2 above)
Comment (gene1wood) [reply functions/federated_aws_rp/app.py:80]: Oh good point

File: functions/federated_aws_rp/app.py (hunk 4 above)
Comment (gene1wood) [reply functions/federated_aws_rp/app.py]: Sure, I think `SameSite=strict` should work as the one case where a user comes to the RP from another site, when they're session needs to be refreshed, the only two pieces of data that the RP uses are `referer` and query parameters

File: functions/federated_aws_rp/app.py (hunk 2 above)
Comment (gene1wood) [reply functions/federated_aws_rp/app.py:80]: Fixed in f98e10db3beaf7d98f4002c0da6a6ec6684814b2

File: functions/federated_aws_rp/app.py (hunk 1 above)
Comment (gene1wood) [reply functions/federated_aws_rp/app.py]: Fixed in f98e10db3beaf7d98f4002c0da6a6ec6684814b2

File: functions/federated_aws_rp/app.py (hunk 4 above)
Comment (gene1wood) [reply functions/federated_aws_rp/app.py]: Fixed in 58eead07041781871f85186762fc4050c9248dd3

File: functions/federated_aws_rp/config.py (hunk 5 above)
Comment (gene1wood) [reply functions/federated_aws_rp/config.py]: Fixed in af102baa0442a35670c755c6eb1f21f6dbc759c4

File: functions/federated_aws_rp/utils.py (hunk 7 above)
Comment (gene1wood) [reply functions/federated_aws_rp/utils.py]: Fixed in f98e10db3beaf7d98f4002c0da6a6ec6684814b2

File: functions/federated_aws_rp/utils.py (hunk 6 above)
Comment (gene1wood) [reply functions/federated_aws_rp/utils.py:74]: Good point. I think I should probably deal with this for all of the requests calls, do something graceful.

File: functions/federated_aws_rp/app.py (hunk 3 above)
Comment (april) [reply functions/federated_aws_rp/app.py:46]: Err, sorry. I meant `.update()`, not `.extend()`.

Comment (gene1wood) [reply functions/federated_aws_rp/app.py:46]: Ah got it. Ya, I think it's fine. If I had a set of new keys to add, `.update()` would be the way to go

File: functions/federated_aws_rp/utils.py (hunk 6 above)
Comment (gene1wood) [reply functions/federated_aws_rp/utils.py:74]: I'll open an issue on this.

**Pull request merged**